engine/table/src/main/java/io/deephaven/engine/table/impl/by/GroupByChunkedOperator.java

```diff
@@ -331,15 +439,123 @@ public ModifiedColumnSet apply(@NotNull final ModifiedColumnSet upstreamModified
+                         ResettableWritableObjectChunk.makeResettableChunk();
+                 final RowSequence.Iterator destinationsIterator =
+                         stepDestinations.getRowSequenceIterator()) {
+
+                // noinspection unchecked
+                final WritableObjectChunk<WritableRowSet, Values> rowSetBackingChunk =
+                        rowSetResettableChunk.asWritableObjectChunk();
+                // noinspection unchecked
+                final WritableObjectChunk<RowSetBuilderRandom, Values> addedBuildersBackingChunk =
+                        addedBuildersResettableChunk.asWritableObjectChunk();
+                // noinspection unchecked
+                final WritableObjectChunk<RowSetBuilderRandom, Values> removedBuildersBackingChunk =
+                        removedBuildersResettableChunk.asWritableObjectChunk();
+
+                while (destinationsIterator.hasMore()) {
+                    final long firstSliceDestination = destinationsIterator.peekNextKey();
+                    final long firstBackingChunkDestination =
+                            rowSets.resetWritableChunkToBackingStore(rowSetResettableChunk, firstSliceDestination);
+                    addedBuilders.resetWritableChunkToBackingStore(addedBuildersResettableChunk, firstSliceDestination);
+                    removedBuilders.resetWritableChunkToBackingStore(removedBuildersResettableChunk, firstSliceDestination);
+
+                    final long lastBackingChunkDestination =
+                            firstBackingChunkDestination + rowSetBackingChunk.size() - 1;
+                    final RowSequence initialDestinationsSlice =
+                            destinationsIterator.getNextRowSequenceThrough(lastBackingChunkDestination);
+
+                    initialDestinationsSlice.forAllRowKeys((final long destination) -> {
+                        final int backingChunkOffset =
+                                Math.toIntExact(destination - firstBackingChunkDestination);
+                        try (final WritableRowSet addRowSet =
+                                nullToEmpty(extractAndClearBuilderRandom(addedBuildersBackingChunk, backingChunkOffset));
+                             final WritableRowSet removeRowSet =
+                                nullToEmpty(extractAndClearBuilderRandom(removedBuildersBackingChunk, backingChunkOffset))) {
+                            final WritableRowSet workingRowSet = rowSetForSlot(backingChunkOffset);
+
+                            // process removes
+                            final boolean wasNonEmpty = removeRowSet.isNonempty();
+                            workingRowSet.remove(removeRowSet);
+                            if (wasNonEmpty && workingRowSet.isEmpty()) {
+                                onEmptied(destination);
+                            }
+
+                            // process adds
+                            final boolean wasEmpty = workingRowSet.isEmpty();
```

Comment:
Moot since I think this is coming out, but I wanted to note that it's *correct* that a state can be emptied and reincarnated on the same cycle, and you are handling that correctly. The helper notes that such rows are "added back" and removes them from adds/removes and instead adds them to modifies.